src/cryptenroll/cryptenroll-fido2.c

```diff
@@ -127,25 +146,61 @@ int enroll_fido2(
         if (keyslot < 0)
                 return log_error_errno(keyslot, "Failed to add new FIDO2 key to %s: %m", node);
 
-        if (asprintf(&keyslot_as_string, "%i", keyslot) < 0)
-                return log_oom();
-
-        r = sd_json_build(&v,
-                       SD_JSON_BUILD_OBJECT(
-                                       SD_JSON_BUILD_PAIR("type", JSON_BUILD_CONST_STRING("systemd-fido2")),
-                                       SD_JSON_BUILD_PAIR("keyslots", SD_JSON_BUILD_ARRAY(SD_JSON_BUILD_STRING(keyslot_as_string))),
-                                       SD_JSON_BUILD_PAIR("fido2-credential", SD_JSON_BUILD_BASE64(cid, cid_size)),
-                                       SD_JSON_BUILD_PAIR("fido2-salt", SD_JSON_BUILD_BASE64(salt, salt_size)),
-                                       SD_JSON_BUILD_PAIR("fido2-rp", JSON_BUILD_CONST_STRING("io.systemd.cryptsetup")),
-                                       SD_JSON_BUILD_PAIR("fido2-clientPin-required", SD_JSON_BUILD_BOOLEAN(FLAGS_SET(lock_with, FIDO2ENROLL_PIN))),
-                                       SD_JSON_BUILD_PAIR("fido2-up-required", SD_JSON_BUILD_BOOLEAN(FLAGS_SET(lock_with, FIDO2ENROLL_UP))),
-                                       SD_JSON_BUILD_PAIR("fido2-uv-required", SD_JSON_BUILD_BOOLEAN(FLAGS_SET(lock_with, FIDO2ENROLL_UV)))));
-        if (r < 0)
-                return log_error_errno(r, "Failed to prepare FIDO2 JSON token object: %m");
-
-        r = cryptsetup_add_token_json(cd, v);
-        if (r < 0)
-                return log_error_errno(r, "Failed to add FIDO2 JSON token to LUKS2 header: %m");
+        if (token_parameters) {
+                if (asprintf(&keyslot_as_string, "%i", keyslot) < 0)
+                        return log_oom();
+
+                r = sd_json_build(&v, SD_JSON_BUILD_OBJECT(
+                                SD_JSON_BUILD_PAIR("type", JSON_BUILD_CONST_STRING("systemd-fido2")),
+                                SD_JSON_BUILD_PAIR("keyslots", SD_JSON_BUILD_ARRAY(SD_JSON_BUILD_STRING(keyslot_as_string))),
+                                SD_JSON_BUILD_PAIR("fido2-credential", SD_JSON_BUILD_BASE64(cid, cid_size)),
+                                SD_JSON_BUILD_PAIR("fido2-salt", SD_JSON_BUILD_BASE64(salt.iov_base, salt.iov_len)),
```

Comment:
use JSON_BUILD_IOVEC_BASE64(salt)

(without the SD_ prefix in this case, we do not intend to export the struct iovec based apis for now)